### ISSUE TITLE
filterService's updateFilters doesn't work with tree data

### ISSUE THREAD

**Issue by s-uds** [ghiscoding/Angular-Slickgrid#754]
### Your Environment

| Software                | Version(s) |
| -------------------- | -------- |
| Angular                  | 8.1.1        | 
| Angular-Slickgrid   | 2.20.1        | 
| TypeScript              | 3.5.3        | 

### Describe the Bug
Hello, I have a custom HTML form that uses `filterService.updateFilters` on columns with `Filters.multipleSelect`, but since I set `enableTreeData: true` in the grid configuration, it doesn't work anymore - oddly it works fine using the header row built-in filter.

### Steps to Reproduce
As described above, manually update the filters using the FilterService with a multipleSelect kind of filter, works just fine until I enabled TreeData in the grid configuration.

### Expected Behavior
`filterService.updateFilters` should work as the header row built-in filter does.

### Current Behavior
The updated filters doesn't seem to apply on the grid data.

### Possible Solution
I've noticed that the DOM filter relies on `callbackSearchEvent`.
Forcing `triggerOnSearchChangeEvent` on `updateFilters` (e.g. `updateFilters(filters, true, true, true)`) so that it calls `callbackSearchEvent` too fixes the issue.

So it might have to do with the way Slickgrid is notified from the filter change?
`updateFilters` uses `this.emitFilterChanged(emitterType);` while `triggerOnSearchChangeEvent` directly uses `this._onSearchChange.notify`.

Thanks a lot!

**Comment by ghiscoding** [comment]
Tree Data is a different beast altogether, it has a few recursive methods (converting from flat to hierarchical dataset) and those needs to be called after a sort, filter and when dataset changes. I'm currently working on fixing a few issues I found while going back on a project with tree data, you can see the current open PR #753 that is in WIP

You could maybe try to `refreshTreeDataFilters` from the Filter Service, I'm not sure if that will work but give it a try and let me know `this.angularGrid.filterService.refreshTreeDataFilters()`

**Comment by s-uds** [comment]
> You could maybe try to refreshTreeDataFilters from the Filter Service, I'm not sure if that will work but give it a try and let me know this.angularGrid.filterService.refreshTreeDataFilters(), if that works I'll add it to the open PR

Yeah that does the trick, thanks a lot!!

> EDIT
Looking back at my comment, I did write the following for Tree Data 😉 and so I'll simply force a trigger it if enableTreeData is set and I'll add to the open PR #753

Ahah, damn, looks like it's missing on the version I use (2.20.1) : p (I mean the triggerOnSearchEvent param isn't documented).

> ahh nevermind I re-read your message and see you just made the last argument true on updateFilters(filters, true, true, true)

Yes, but actually I also had to comment out `this.updateColumnFilters(newFilter.searchTerms, uiFilter.columnDef, newOperator)` in [updateFilters](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/modules/angular-slickgrid/services/filter.service.ts#L788) otherwise it wouldn't pass the `isequal(oldColumnFilters, this._columnFilters)` in [callbackSearchEvent](https://github.com/ghiscoding/Angular-Slickgrid/blob/de8fbcc49722cc9790b9ae5f01bd72f872350b7d/src/app/modules/angular-slickgrid/services/filter.service.ts#L976).

**Comment by ghiscoding** [comment]
I did the change already in the 3c603a7 commit and it works, for now on your side you can just call the last argument as True and that should work. I made the change in the commit so that if it detects that it's a Tree Data then it will trigger that onSearch callback that you found out about.

So basically the following 2 lines will be equivalent when using Tree Data
```ts
this.angularGrid.filterService.updateFilters([yourFilters], true, true, true);

// with the new commit, this will also do the exact same thing
this.angularGrid.filterService.updateFilters([yourFilters]);
```

I have another thing to look into and add unit tests, I should have a new version by the end of the week. You're actually lucky because I starting working on this last Friday 😉 So it's a good timing, the PR will include a bunch of other fixes as well, you can see the list of fixes in the PR #753

**Comment by s-uds** [comment]
Cool, Thanks again! Although like I said in the previous post, for me that extra argument wasn't enough:

```ts
// in FilterService.updateFilters

filters.forEach((newFilter) => {
  const uiFilter = this._filtersMetadata.find((filter) => newFilter.columnId === filter.columnDef.id);
  if (newFilter && uiFilter) {
      const newOperator = newFilter.operator || uiFilter.defaultOperator;
      
     /**
      * I had to comment out the line below, otherwise isequal(oldColumnFilters, this._columnFilters) would return true, in callbackSearchEvent()
      */
      // this.updateColumnFilters(newFilter.searchTerms, uiFilter.columnDef, newOperator);
      
      uiFilter.setValues(newFilter.searchTerms, newOperator);
      if (triggerOnSearchChangeEvent) {
          this.callbackSearchEvent(null, { columnDef: uiFilter.columnDef, operator: newOperator, searchTerms: newFilter.searchTerms, shouldTriggerQuery: true });
      }
  }
  
  // (...)
});

```

So I'm gonna stick with `this.angularGrid.filterService.refreshTreeDataFilters()` as of now

**Comment by ghiscoding** [comment]
> Cool, Thanks again! Although like I said in the previous post, for me that extra argument wasn't enough:

right, I actually found that problem when writing the unit test, I added an internal flag `forceOnSearchChangeEvent` to fix that in the commit I referenced earlier on this [line](https://github.com/ghiscoding/Angular-Slickgrid/commit/3c603a73bf08674c4de33fc4cd93dba5e6570292#diff-4e06dd7bdaa67e6c3ff2434d0928c42459e9bb8a41efef7faf939ff558128271R792), so in the `updateFilters()` it will set that flag to true and it will work. You can use `refreshTreeDataFilters` for now but I would suggest to change it back once the new version is out. 

I'm still working on this and was facing some racing conditions but I think that I'm finally near the end of it. The new version should improve performances as well since I found that the grid was being rendered and resorted multiple times because of these racing conditions and other factors. 

BTW, I'm just curious which type of Tree Data are you using? with Parent/Child relationship or a dataset that is already a hierarchical tree structure?

**Comment by s-uds** [comment]
> right, I actually found that problem when writing the unit test, I added an internal flag forceOnSearchChangeEvent to fix that in the commit I referenced earlier on this line, so in the updateFilters() it will set that flag to true and it will work.

neat!

> BTW, I'm just curious which type of Tree Data are you using? with Parent/Child relationship or a dataset that is already a hierarchical tree structure?

Just a parent/child relationship, I add new children rows dynamically - but actually I'm facing another issue as I realize that it doesn't work (dynamic child rows) after using (and removing) a filter at least once..

**Comment by ghiscoding** [comment]
we also use parent/child relationship in all 3 grids we are starting development with, I think most people will do that as well.

> Just a parent/child relationship, I add new children rows dynamically - but actually I'm facing another issue as I realize that it doesn't work (dynamic child rows) after using (and removing) a filter at least once..

do you mean updating the filter dynamically every time? If that is the case, then you might need to wait for my next version, current code seems to be working good as you can see in the animated gif. I was using a fixed number at first, but just to show you that it works dynamically, I made the "% complete" smaller than a random number. Was that your problem?

![XOetkbJbl7](https://user-images.githubusercontent.com/643976/117055571-4df21400-ace9-11eb-9d92-a045d35b7d4e.gif)

**Comment by s-uds** [comment]
> do you mean updating the filter dynamically every time?

No, I add/remove child rows dynamically using `this.angularGrid.dataView.addItem(newItem);` which works just fine until I use a (dynamic) filter:

```ts
// works unless I used (and removed) filters dynamically at least once
addChildRows() {
     this.angularGrid.dataView.addItem(newItem);

     const dataset = this.angularGrid.dataView.getItems();
     this.dataset = [...dataset]; // make a copy to trigger a dataset refresh
     
     setTimeout(() => {
           const idColumn = this.columnDefinitions.find(col => col.id === 'id') as Column;
           this.angularGrid.sortService.onLocalSortChanged(this.slickGrid, [{ columnId: 'id', sortCol: idColumn, sortAsc: true }]);
     }, 0);
}

// even after doing this, the new rows don't show up
removeFilters() {
      this.angularGrid.filterService.clearFilters();
      this.angularGrid.filterService.refreshTreeDataFilters();
      this.angularGrid.gridService.resetGrid(this.columnDefinitions);
}
```

I hope it's clear, basically the "add dynamic rows" and "add dynamic filters" features don't coexist (I use either one or the other) but for some reason, using "add dynamic filters" breaks the other one.

**Comment by ghiscoding** [comment]
I did also see that bug and and fixed the Add Item, you'll have to change your code though because I fixed it by doing necessary calls in the Grid Service. 
```ts
this.angularGrid.gridService.addItem(newItem);
```
This way is much simpler since that service is one I've done and have control (I don't have control on the one from dataview because that comes from the core lib). 

Here's the new code from my example. 
![image](https://user-images.githubusercontent.com/643976/117140371-585cee00-ad7b-11eb-8da7-8400b9e10b1b.png)
![SVjOhMtpHk](https://user-images.githubusercontent.com/643976/117140594-91955e00-ad7b-11eb-8dd7-d2e231b99699.gif)

**Comment by s-uds** [comment]
Interesting, thanks - from which version is that fix available?

I believe I'm using the latest version that still supports Angular 8 (**2.20.1**)

**Comment by ghiscoding** [comment]
it's not yet available, I didn't merge the PR #753 yet 😝

**Comment by s-uds** [comment]
Ah I see!

I think I've pinpointed the issue:
After using filters (and **only with**), `dataView` keeps some sort of a reference for all (pre) existing rows ids, and only those are taken into account - so trying to insert a row with a new id just won't work (even with a new dataset).

Using `this.angularGrid.dataView.destroy();` solves the problem, but obviously is quite radical and resource consuming..

Do you know if there's any other way I could quickly reset that dataView id reference made from using filters?

Thanks again

**Comment by ghiscoding** [comment]
So as I mentioned in my first reply on top, Tree Data is very different beast. You can't just play with the data and DataView and expect it to work. For the Tree Data to work, it must be sorted every time and recalculate each element tree level indentation and for that sort to work it must be converted from a flat data structure to a hierarchical (tree) structure, then sort it and then reconvert to a flat data structure. If you play with the DataView directly then you'll bypass any kind of sorting and it will probably also fall outside of the correct tree level indentation. 

In the version I'm working now, I added some more functions that could be used from the outside as a user, but you'll have to wait. I think it's better if you just wait for the new version, unless you want to go and call yourself the `convertParentChildArrayToHierarchicalView` and `convertHierarchicalViewToParentChildArray`. There's a lot of code I had to put in place for all of Tree Data to work.. I'd suggest you wait until it's ready. 

I'm doing yet another thing, the Export wasn't exporting with the indentation, I'm adding a new TreeExportFormatter. I should be done today and then perhaps I can ship a version tomorrow. in the demo below, I had to add a dot "." in front of an indented text because it seems Excel is trimming white space when it's at beginning, oh well

![image](https://user-images.githubusercontent.com/643976/117158373-5a2fad00-ad8d-11eb-8a52-3edcf262821b.png)

**Comment by ghiscoding** [comment]
@s-uds hey I don't think that I will push a new version this week, most probably next week instead. If you can't really wait then I'd suggest cloning the repo (since the PR got merged), then install node modules and run `npm run build`, then take the content of the `dist` folder and paste that in your `node_modules/angular-slickgrid` to overwrite its content... or simply wait next week, up to you 😉

**Comment by s-uds** [comment]
Hey, alright thanks for the heads up - I'll have to go through updating Angular first which has yet to be planned and accounted, but once that is done, I'll make sure to grab the latest version you provided.

I've finally managed to make it work with a temporary trick: I just add and remove an empty filter to force tree data update (because `refreshTreeDataFilters` wasn't enough) and then dataView.addItem is effective again.

Btw is there plans to have an `insertItem(before, item)` method in the Grid service, similar to the one in dataView?

**Comment by ghiscoding** [comment]
> Btw is there plans to have an `insertItem(before, item)` method in the Grid service, similar to the one in dataView?

no plans, the `addItem` has a `position` option for top/bottom but that's all and that is always enough for me. You have 2 ways to deal with this
1. use the DataView directly
    - the Grid Service is more of a helper, I created it to avoid have to write the same few lines of code over and over but you can still go with the DataView directly
2. contribute to the project and maybe add a `positionIndex` that can be an extra option to the `addItem`

**Comment by ghiscoding** [comment]
the PR #759 for Angular-Slickgrid is in, I'm not sure when I'll release though, it might be this week or next week. There is still 1 other thing I want to look into and it might take some time

**Comment by ghiscoding** [comment]
It took me longer than I expected but the new version is now out and there's a few other features as well, so go check it out.
Take a look at the new version [2.29.x](https://github.com/ghiscoding/Angular-Slickgrid/releases/tag/v2.29.0)

Cheers ⭐
